rspace/src/main/scala/coop/rchain/rspace/history/package.scala

```diff
@@ -0,0 +1,191 @@
+package coop.rchain.rspace
+
+import java.lang.{Byte => JByte}
+
+import com.typesafe.scalalogging.Logger
+import coop.rchain.shared.AttemptOps._
+import scodec.Codec
+
+import scala.annotation.tailrec
+
+package object history {
+
+  private val logger: Logger = Logger[this.type]
+
+  def initialize[T, K, V](store: ITrieStore[T, K, V])(implicit
+                                                      codecK: Codec[K],
+                                                      codecV: Codec[V]): Unit = {
+    val root     = Trie.create[K, V]()
+    val rootHash = Trie.hash(root)
+    store.withTxn(store.createTxnWrite()) { txn =>
+      store.put(txn, rootHash, root)
+    }
+    store.workingRootHash.put(rootHash)
+    logger.debug(s"workingRootHash: ${store.workingRootHash.get}")
+  }
+
+  def lookup[T, K, V](store: ITrieStore[T, K, V], key: K)(implicit codecK: Codec[K]): Option[V] = {
+    val path = codecK.encode(key).map(_.bytes.toSeq).get
+    @tailrec
+    def loop(txn: T, depth: Int, curr: Trie[K, V]): Option[V] =
+      curr match {
+        case Node(pointerBlock) =>
+          val index: Int = JByte.toUnsignedInt(path(depth))
+          // We use an explicit match here instead of flatMapping in order to make this function
+          // tail-recursive
+          pointerBlock.toVector(index) match {
+            case None =>
+              None
+            case Some(hash: Blake2b256Hash) =>
+              store.get(txn, hash) match {
+                case Some(next) => loop(txn, depth + 1, next)
+                case None       => throw new LookupException(s"No node at $hash")
+              }
+          }
+        case Leaf(lk, lv) if key == lk =>
+          Some(lv)
+        case Leaf(_, _) =>
+          None
+      }
+    store.withTxn(store.createTxnRead()) { (txn: T) =>
+      store.get(txn, store.workingRootHash.get).flatMap { (currentRoot: Trie[K, V]) =>
+        loop(txn, 0, currentRoot)
+      }
+    }
+  }
+
+  @tailrec
+  private[rspace] def getParents[T, K, V](
+      store: ITrieStore[T, K, V],
+      txn: T,
```

Comment:
Transaction is our context, so I'd move it to the implicit params list so there's less code on the caller side

Reply:
This is an interesting idea that I had not yet considered.  Let's talk about the pros and cons of refactoring along these lines together as a team next week.